common/modules/reserve_queue/src/reserve_queue.rs

```diff
@@ -0,0 +1,417 @@
+#![no_std]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+const DEFAULT_PRECISION: u64 = 1000;
+
+mod reserve_node;
+use reserve_node::{Nonce, NonceAmountPair, ReserveNode};
+
+#[elrond_wasm::module]
+pub trait ReserveQueueModule {
+    #[endpoint]
+    fn pop(
+        &self,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+    ) -> SCResult<Vec<NonceAmountPair<Self::BigUint>>> {
+        let mut result = Vec::new();
+        let big_zero = Self::BigUint::zero();
+        let mut amount_left = amount.clone();
+        let queues_orig = self.queues(token_id).get();
+        let mut queues = queues_orig.clone();
+        let precision = self.get_precision_or_default(token_id);
+
+        while amount_left != big_zero && !queues.is_empty() {
+            let queue_id = *queues.last().unwrap();
+            let mut queue_result =
+                self.pop_from_queue(token_id, &amount_left, precision, &mut queues, queue_id)?;
+
+            let mut queue_result_sum_amounts = big_zero.clone();
+            queue_result
+                .iter()
+                .for_each(|x| queue_result_sum_amounts += &x.amount);
+
+            result.append(&mut queue_result);
+            amount_left -= &queue_result_sum_amounts;
+        }
+        require!(amount_left == big_zero, "Not enough reserves");
+
+        if queues != queues_orig {
+            self.queues(token_id).set(&queues);
+        }
+
+        Ok(result)
+    }
+
+    #[endpoint]
+    fn push(
+        &self,
+        token_id: &TokenIdentifier,
+        pairs: &[NonceAmountPair<Self::BigUint>],
+    ) -> SCResult<()> {
+        let queues_orig = self.queues(token_id).get();
+        let mut queues = queues_orig.clone();
+        let precision = self.get_precision_or_default(token_id);
+        let pairs_compressed = self.get_compressed_pairs(pairs);
+
+        for pair in pairs_compressed.iter() {
+            self.push_single(token_id, pair.nonce, &pair.amount, precision, &mut queues)?;
+        }
+
+        if queues != queues_orig {
+            self.queues(token_id).set(&queues);
+        }
+
+        Ok(())
+    }
+
+    // Below funcs should be private
+
+    fn pop_from_queue(
+        &self,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+        queue_id: u64,
+    ) -> SCResult<Vec<NonceAmountPair<Self::BigUint>>> {
+        let is_fungible = !self.reserve(token_id, 0).is_empty();
+
+        if is_fungible {
```

Comment:
do we actually need this if's and else's ? the queue should have actually only 1 element with fungible tokens. Which might still happen on Semi Fungible that everyone will put only 1 type of token inside - like one nonce ?

Reply:
I used nonce 0 in order to know whether the head/tail elements are instantiated or not. Similar to NULL. May refactor this, but some magic value will be used for this check anyway